### DISCUSSION TITLE
Add an entity decorator to handle user permisionning

### DISCUSSION THREAD

**Discussion by BrannJoly** [mikro-orm/mikro-orm#5767] [Ideas]
**Is your feature request related to a problem? Please describe.**

The current migration generation works great, but it doesn't handle at all the grants part of the schema.

This is fine in dev as long as you use a database admin user (eg 'postgres') to test your backend, but a more relatistic setup in production would involve using a less privileged database user to access the database.
I didn't find a built in solution to handle this scenario.

**Describe the solution you'd like**
I'd like an entity decorator :
``` javascript
@Grant({user : 'backend_user', select: true, insert: true, update: true, delete: false}
@Entity({ tableName: 'table_name' })
```
This would be picked up by the migrator to generate the appropriate code eg for postgres
``` postgresql
GRANT SELECT, INSERT, UPDATE ON TABLE table_name TO backend_user;
```

**Describe alternatives you've considered**

I considered using a [custom migration generator](https://mikro-orm.io/docs/migrations#using-custom-migrationgenerator)
This would involve overriding createStatement, parsing the sql provided, check if it is a table creation statement, and if so appending the appropriate sql to grant the correct permissions.
This would be error-prone (parsing raw sql ...), and also this wouldn't allow for fine-grained control on a per table basis

**Comment by B4nan** [comment]
> but a more relatistic setup in production would involve using a less privileged database user to access the database.

Isn't this the exact reason this wouldn't work? Or you are using a different (more privileged) user to execute the migrations than the one to run the app?

I checked if something like this exists in https://hibernate.org/orm/ or doctrine and I can't see they would deal with this anyhow. It feels quite weird to manage permissions from inside migrations to me.

**Reply by BrannJoly** [reply]
> Isn't this the exact reason this wouldn't work? Or you are using a different (more privileged) user to execute the migrations than the one to run the app?

Yes, the migrations are executed with a login that has schema update privileges
and then my backend itself runs with a less privileged user that can't touch the schema (and has limited access to some tables as well)

It seems to me that I have the following options:
- use a privileged user to run my backend (sounds like a bad idea to me)
- write all the GRANT code manually (sounds like a bad idea to me)
- having my migration tool generate the GRANT code on top of the CREATE TABLE code (sounds like a bad idea to you :p)

> It feels quite weird to manage permissions from inside migrations to me.

When a new tenant (hence a new db) is provisionned on my app, I use migrations to set up everything (including required db extensions, etc). So it seems like a logical place to handle users and grants as well. Actually I can't see another option unless you're using umbrella permissions i.e. `GRANT ALL PRIVILEGES ON database_name.* TO user`

**Reply by B4nan** [reply]
The reason why I don't like this idea is that I don't know how schema diffing for it could work. You need to be able to tell whether the query is already executed or not based on the existing schema (not just the snapshot file, that would be rather simple). But I am not completely opposed to it either.

If you could provide more details on how this can work (at least for MySQL and PostgreSQL), I guess I can handle the rest.

**Reply by BrannJoly** [reply]
On postgres at least, you have all the information available in table `information_schema.role_table_grants`
Or did I misunderstood your question?
